### ISSUE TITLE
Design Meeting Notes, 8/2/2023

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#55244]
# Isolated Declarations

#53463

* If you have a syntactic declaration emitter, you can build every file in parallel.
* [See chart](https://github.com/microsoft/TypeScript/pull/53463#issuecomment-1660720127) of series versus topological-parallel versus fully parallel.
* Last time we talked about this we said "can't TypeScript do nothing?" but each implementer is looking to us for an answer.
    * Implementers want a common spec.
    * But also, TypeScript needs to be able to say "an isolated declaration emitter *would not* create a *correct* declaration file".
* Some of these seem reasonable - but you start to get into more and more complex examples that are questionable.
* Why is this TypeScript's job? TypeScript might not be able to know what esbuild can do versus swc versus...
* TypeScript could do a few different levels:
    * Create a set of rules that must be followed for isolated declaration emit, make sure that code conforms to it.
    * Checking of non-trivial inferences.
* `const enum` inlining in esbuild is a great counter-example of why 1 flag to control this all doesn't necessarily scale.
* People seem unsettled by the `satisfies` example in the write-up.
* There's a version of this that just says "everything's gotta be annotated, end-of-story".
* Could diff between TS `.d.ts` emit and external tool `.d.ts` emit?
    * `string`-based comparison?
    * Each project could still do this check independently of any other project.
* Feels like the idea of "everything should be annotated" feels better than speculation?
    * What about object literals with `>400` fields or something?
    * Well, we don't feel good about it...
    * Happens with big CSS objects that are shared across files, options/configs, etc.
* This targets declaration emit at every file or every project?
    * Could be either really.
* Under this mode, circular project references no longer become a problem.
* One idea: start conservative and tell users that if they want to draw outside the lines, they can turn off `isolatedDeclarations` and build a tool to enforce the special cases. Larger orgs are more well-equipped to do so.
* Another idea: ship something, allow more later. 3rd-party compilers can say "I don't understand that" later.
* Both of these ideas actually should work out fine.
* Feels weird that the people championing this wouldn't use this flag!
    * tsc would be utterly irrelevant for their `.d.ts` and `.js` build pipelines - but is still a key part of their workloads.
* How does this work for decorator type mutations?
    * We'd figure it out.
* So what would we propose.
    * We don't like:
        * INF04: `Symbol`
        * INF05: `typeof`
        * INF09: `new` operator
        * INF10: `<jsx>`
        * INF15: return type inference
        * INF16: conditionals (object normalization, subtype reduction)
    * We need to review:
        * INF07: `as const` whether it's `= 1` vs `: 1`
        * INF11: `typeof` in the example seems like a no.
        * INF14: Object normalization and subtype reduction.
        * INF17: `satisfies` - depends on whether or not `const`
* Got as far as INF17, out of time.

**Comment by DanielRosenwasser** [comment]
@dragomirtitian, @robpalmer, @h-joo

**Comment by DanielRosenwasser** [comment]
Continuing on feedback from https://github.com/bloomberg/TypeScript/blob/isolated-declarations/external-declarations/docs/local-inferences.md

* *I* don't like (and others like @RyanCavanaugh can nitpick on this)
    * INF20: Object spreads.
    * INF22: Array spreads in non-const contexts.
* Review
    * INF18: Template literals in non-const contexts. Feels right. We *did* try inferring literals a while back, but kind of failed at it.
    * INF19: Unclear about anything advanced.
    * INF21: Array spreads: See `typeof` feedback above. Is it okay to use `typeof` on something `as const`? Is that hard for compilers to track?
    * INF23: Expando properties on functions.

**Comment by DanielRosenwasser** [comment]
* INF01: Yes - but the example emit is wrong.
    * ```ts
      const x = 123;
      const y: 123 = 123;

      let a = x;
      let b = y;

      // becomes...

      declare const x = 123;
      declare const y: 123;
      declare let a: number;
      declare let b: 123;
      ```
* INF02: Yes - but this is `--strictNullChecks`-sensitive. So `isolatedDeclarations` might mandate that!
* INF03: Yes - but this is `--strictNullChecks`-sensitive. So `isolatedDeclarations` might mandate that!
* INF04: No - let's revisit after shipping. Also, note that TypeScript actually shadows `Symbol`. May be a challenge for some compilers.
* INF05: No - narrowing affects values, but `typeof` captures the declared type instead of the narrowed type. Also not what our declaration emit currently gives.
* INF06: Yes. Asserted type always flows out.
* INF07: Yes (no longer under review!). See notes about widening vs. non-widening in INF01 though. Currently these are the non-widening.
* INF08: Yes. As long as they're annotated. Aside: the rule about fully annotated functions - pretty close to `noImplicitAny`. Maybe we say `--strict` is mandated? Let's not rabbit-hole too much...
* INF09: No - let's revisit after shipping. Aside: if you ever wanted to support this, the following would be a good test case. You'd want it to work.

  ```ts
  // a.ts
  export declare const Dog: new () => Cat;
  export interface Cat {}

  // b.ts
  import { Dog as DogCtor, Cat } from "./a";

  export const Dog = DogCtor;
  export type Dog = Cat;

  // c.ts
  import { Dog } from "./b";

  let d = new Dog();
  ```
* INF10: No.
* INF11: This is fine, but has to fall out of existing rules.

  Aside, and back to `typeof`: occasionally you will need to emit a local for declaration emit outside of `export default`. [See this example](https://www.typescriptlang.org/play#code/KYDwDg9gTgLgBAbwL4CgUHp1wCoTgE2AGMBLQuEgMzhgAsSBnCpgG2AHMBDFgGjgE8IAVzgA7YMHw08nUQwDuwKDVrA4ARyHAGMEhDlwARsBYR5KUJFhwi+nXEsAuGvzDAI1TnAC8iOCGcABj5+ZyFRQkoScSkkAG40TDgARgA6OAARCG0VNXwYJmAAWxIYGCU4AGtRM2kHEvgvUyJuOEpoFUYAfgwsACZ0rJy6PIL60vLlatq6RgJiFk4oTl19ZjgGIUMAK2J4GBlRGpgVvVF0aMolYFEiNSghNgYe2zlGnz8AlMCfkOcAWRWtFSywiECKAAoAJRwLpwADK-CKhggLGhcDCEWAURicFQKCAA) that @andrewbranch constructed:

   ```ts
   export {}

   // To decide if this is legal, you need to answer the questions below
   export const exp: typeof a = { x: 0, y: undefined };

   // 1. Does the dts emitter know to emit a local for this?
   // 2. Does the dts emitter know this declaration is subject to annotation/inference rules?
   const a = { x: 1000, y: Math.random() ? Symbol() : undefined }
   ```

   A declaration emitter would need to be able to resolve `a` when emitting `exp`, run the declaration emit inference rules over it, and keep track that `a` must be preserved.

   Also, if we ever decided to do a speculative/inferential mode, checking for differences between the declared/narrowed type for use in `typeof` might be fair game. But there may also be cases where `typeof` doesn't truly substitute appropriately. We can't think of a specific example, but the idea is in the same vein as answering whether or not to emit the first or second version of `x` in the following example:

   ```ts
   export interface MergableType {
       abc: number;
   }
   export let x: MergableType[keyof MergableType];
   // vs.
   export let x: number;
   ```

   Interface merging, non-accessible locals start to make this funky. That's why starting conservative here might be the best so we can think through these more thoroughly.
* INF12: Yes, see notes on INF11 though. No spreads.
* INF13: Yes, see notes on INF11 though. No spreads.
* INF14: No. Object normalization and subtype reduction.
* INF15: No. Object normalization and subtype reduction.
* INF16: No. Object normalization and subtype reduction. Aside: we could image compilers permitting a pass where if each branch is *identical*. But we can revisit that at a later time.
* INF17: No. Would make sense if we had Flow's `(expr : Type)` syntax or something. Something to consider for property types on objects?
* INF18: Yes. Always string.
* INF19: **Uhhhhhh.** ❓ We're not sure. We feel mixed about the emitter leaning on the checker to do the actual subtype reduction.
* INF20: No. Really relies on `typeof`, and is lossy.
* INF21: No. Narrowing.

  ```ts
  function foo() {
      return 0;
  }

  declare function wat(x: any): asserts x is readonly [0];

  const arr = [foo()] as const;

  wat(arr);

  export const arr2 = [...arr] as const;
  ```
* INF22: No. Object normalization and subtype reduction.
* INF23: No. Revisit.
* Didn't get to review open issues yet.

**Comment by fatcerberus** [comment]
I'm lost here - what do these `INFxx` designations refer to?  Usually it's nice to have the meeting notes public, but this one seems impenetrable.

**Comment by jakebailey** [comment]
See: https://github.com/bloomberg/TypeScript/blob/isolated-declarations/external-declarations/docs/local-inferences.md

**Comment by DanielRosenwasser** [comment]
Something that got brought up about initializers:

```ts
export declare let { x = "hello", y }: { x: string | undefined, y: string };
```

Something people are thinking about is just saying that binding patterns are not allowed - at least at first.
